lib/streamlit/state/widgets.py

```diff
@@ -80,15 +84,15 @@ class NoValue:
 def register_widget(
     element_type: str,
     element_proto: WidgetProto,
-    deserializer: WidgetDeserializer,
-    serializer: WidgetSerializer,
+    deserializer: WidgetDeserializer[T],
+    serializer: WidgetSerializer[T],
     ctx: Optional["ScriptRunContext"],
     user_key: Optional[str] = None,
     widget_func_name: Optional[str] = None,
     on_change_handler: Optional[WidgetCallback] = None,
     args: Optional[WidgetArgs] = None,
     kwargs: Optional[WidgetKwargs] = None,
-) -> Tuple[Any, bool]:
+) -> Union[DisconnectedReport, WidgetStateReport[T]]:
```

Comment:
Have a look at the last line of `register_widget`:
```return ctx.session_state.register_widget(metadata, user_key)```

This refers to `SafeSessionState.register_widget`, which does can do the following:
```py
with self._lock:
    if self._disconnected:
        return DisconnectedReport()
```

So, `register_widget` can indeed return a `DisconnectedReport`.

I created a separation between `DisconnectedReport` and `WidgetStateReport`, because the latter is generic with regards to T, but the former cannot be. I also found it useful to let the class name indicate the cause of the report. But, I am realising that `SafeSessionState.register_widget` should probably just return an _optional_ `WidgetStateReport` instead. I think that would be less confusing.

Here's a question I hope you can answer for me, though. Is there a good reason for the difference in fallback logic for the  `ctx is None` case in `register_widget`, and the `self._disconnected` case of `SafeSessionState.register_widget`? I was really tempted to simplify all this by using the logic of `register_widget` in both cases, but I was afraid of changing behaviour too much. Am now realising one could just return `None` from `register_widget` also, having the return type be `Optional[WidgetStateReport[T]]`, with the `None` case covering both disconnection and missing `ctx`. I think this would be my preferred solution, unless you see something I don't. It would also realign this function with what the docstring claims it does.

This does, however, put the responsibility of figuring out how to fall back in the case or a missing  `WidgetStateReport[T]` firmly on the widget calling `register_widget`. That's not a bad thing, but it probably means someone other than me needs to properly review whether that fallback logic in each individual widget is sound, because I don't intimately know what your specs say should happen in each case.